### DISCUSSION TITLE
How long brew bottles (precompiled binaries ) are kept?

### DISCUSSION THREAD

**Discussion by vroad** [Homebrew/discussions#1183] [Everyday usage] [answered]
I've read somewhere that brew bottles are not kept forever, but forgot where...
That's one of the reason why brew doesn't let you pin package versions easily with something like npm's package-lock.json, if I remember correctly.
Is there any official documentation for retention policy, or comments from homebrew development team on the issue?

**Comment by gromgit** [comment, marked as answer]
Not a maintainer, but it's generally easy to predict when a formula's bottle for macOS version _X_ will fall out of general availability, since Homebrew support basically [tracks macOS support](https://en.wikipedia.org/wiki/MacOS_version_history#Releases), so as of this writing, only three macOS versions are supported (Mojave, Catalina, and Big Sur).

Whenever a formula is _version-bumped_, only bottles for the macOS versions supported _at the time of the bump_ will be built, simply because there' s no Homebrew hardware to build for older versions. If it's _not_, then whenever a new macOS version is released, a bottle will be built and _added_ to the existing list.

So if a software project is in active development, the corresponding formula will generally only have bottles for the last three versions. At the other end of the spectrum, "stale" projects like [globe](https://www.acme.com/software/globe/) still have bottles stretching back to Mountain Lion: https://formulae.brew.sh/formula/globe

When a specific bottle is actually _deleted_, i.e. is not available even if you `brew extract` an earlier version of a formula, is a Homebrew maintainer decision.

**Reply by vroad** [reply]
I understood that latest versions are only supported for recent macOS versions.

> When a specific bottle is actually deleted, i.e. is not available even if you brew extract an earlier version of a formula, is a Homebrew maintainer decision.

Is it deleted because keeping all versions costs too much? Or are the any other reasons?

**Reply by SMillerDev** [reply]
> Is it deleted because keeping all versions costs too much? Or are the any other reasons?

We only ever support the latest version of software so any reason would be enough to remove old bottles. 

So far we haven't removed any that I know of, but with the upcoming migration away from bintray we might only migrate latest versions for example.
